### DISCUSSION TITLE
SHACL validation not considering datatypes

### DISCUSSION THREAD

**Discussion by MajewskiKrzysztof** [eclipse-rdf4j/rdf4j#4144] [Q&A]
I've noticed an invalid behavior when running SHACL validation for JSON-LD input.
This is my code:
```java
String SHAPES = "C:\\Users\\MAJEWSK6\\Desktop\\SHACL_TEST\\rulesJsonLd.jsonld";
String DATA = "C:\\Users\\MAJEWSK6\\Desktop\\SHACL_TEST\\inputJsonLd.jsonld";
​
ShaclSail shaclSail = new ShaclSail(new MemoryStore());
Repository repo = new SailRepository(shaclSail);
​
try (RepositoryConnection connection = repo.getConnection()) {
    connection.begin();
    connection.add(new StringReader(Files.readString(Path.of(SHAPES))), RDFFormat.JSONLD, RDF4J.SHACL_SHAPE_GRAPH);
    connection.commit();
​
    connection.begin();
    connection.add(new StringReader(Files.readString(Path.of(DATA))), RDFFormat.JSONLD);
    connection.commit();
​
    connection.begin();
    connection.clear(RDF4J.SHACL_SHAPE_GRAPH);
    connection.commit();
} catch (Exception e) {
    Throwable cause = e.getCause();
    if (cause instanceof ValidationException) {
        Model validationReportModel = ((ValidationException) cause).validationReportAsModel();
        Rio.write(validationReportModel, System.out, RDFFormat.TURTLE);
    }
}
```
The payload file:
```json
{
    "@context": {
        "Class": "http://example.com/Class",
        "property": {
            "@id": "http://example.com/Property",
            "@type": "http://www.w3.org/2001/XMLSchema#integer"
        }
    },
    "@graph": {
        "@type": "Class",
        "@id": "http://example.com/1",
        "property": "test"
    }
}
```
And the rules file:
```json
{
  "@context": {
    "@base": "http://ontology.roche.com/",
    "@vocab": "http://ontology.roche.com/model/rts/core/",
    "id": "@id",
    "type": "@type",
    "datatype": {
      "@id": "http://www.w3.org/ns/shacl#datatype",
      "@type": "@id"
    },
    "path": {
      "@id": "http://www.w3.org/ns/shacl#path",
      "@type": "@id"
    },
    "class": {
      "@id": "http://www.w3.org/ns/shacl#class",
      "@type": "@id"
    },
    "describes": {
      "@type": "@id"
    },
    "property": {
      "@id": "http://www.w3.org/ns/shacl#property",
      "@type": "@id",
      "@container": "@set"
    },
    "targetClass": {
      "@id": "http://www.w3.org/ns/shacl#targetClass",
      "@type": "@id"
    },
    "maxCount": "http://www.w3.org/ns/shacl#maxCount",
    "minCount": "http://www.w3.org/ns/shacl#minCount",
    "PropertyShape": "http://www.w3.org/ns/shacl#PropertyShape",
    "Property": "http://www.w3.org/1999/02/22-rdf-syntax-ns#Property",
    "NodeShape": "http://www.w3.org/ns/shacl#NodeShape",
    "Class": "http://www.w3.org/2002/07/owl#Class",
    "DatatypeProperty": "http://www.w3.org/2002/07/owl#DatatypeProperty"
  },
  "@graph": [
    {
      "id": "http://example.com/Property",
      "type": [
        "DatatypeProperty",
        "Property"
      ]
    },
    {
      "id": "http://example.com/Class",
      "type": "Class"
    },
    {
      "id": "http://example.com/ClassShape",
      "type": "NodeShape",
      "property": [
        "http://example.com/PropertyShape"
      ],
      "targetClass": "http://example.com/Class"
    },
    {
      "id": "http://example.com/PropertyShape",
      "type": "PropertyShape",
      "datatype": "http://www.w3.org/2001/XMLSchema#integer",
      "maxCount": 1,
      "minCount": 1,
      "path": "http://example.com/Property"
    }
  ]
}
```
​
After running this example I get no violations.
Shouldn't the "test" cause `shacl:DatatypeConstraintComponent` when passed as integer?
​
The same issue exists for other datatypes like `xsd:dateTime` for example. It accepts any string value.

**Comment by hmottestad** [comment]
When I convert your data to NQUADS it looks like this:
```
<http://example.com/1> <http://example.com/Property> "test"^^<http://www.w3.org/2001/XMLSchema#integer> .
<http://example.com/1> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://example.com/Class> .
```

It looks like you are expecting the SHACL validation to check that the value conforms to the restrictions of the datatype. My understanding of the SHACL specification is that we should rely on the specified datatype and not perform any other checks if the datatype matches the constraint.

RDF4J supports checking that the value conforms to the datatype during parsing. See https://rdf4j.org/javadoc/latest/org/eclipse/rdf4j/rio/helpers/BasicParserSettings.html#VERIFY_DATATYPE_VALUES and https://rdf4j.org/documentation/programming/rio/#configuring-the-parser--writer

**Reply by MajewskiKrzysztof** [reply]
Hi @hmottestad,
thanks a lot for your answer.
I've added this configuration to my code:
```java
connection.getParserConfig().set(BasicParserSettings.VERIFY_DATATYPE_VALUES, true);
```
and I'm getting proper datatype format validation:
```
org.eclipse.rdf4j.rio.RDFParseException: 'test' is not a valid value for datatype http://www.w3.org/2001/XMLSchema#dateTime
```
but it's returned as a Java exception and not SHACL violation.
Other SHACL engines like Apache Jena, for example, report these kind of things as a SHACL violations. It this their own interpretation of SHACL?
Or is there any other way to use RDF4J and have this errors as SHACL violations?

**Reply by hmottestad** [reply]
Funky. I tried the SHACL playground and it kinda works. If I use your example it gives a validation error. If I change it to use the string "1" with xsd:integer as datatype it passes validation, but using "0.5" also passes validation which should not be the case.

**Reply by hmottestad** [reply]
I also tried switching from xsd:integer to xsd:int there doesn’t seem to any validation and the "test" string with xsd:int as datatype passes validation.

**Reply by hmottestad** [reply]
You could post to the SHACL mailing list and ask what the recommended approach is.

**Reply by afs** [reply]
https://www.w3.org/TR/shacl/#DatatypeConstraintComponent
There is an "ill-typed" condition for some XSD datatypes. 

> A [literal](https://www.w3.org/TR/shacl/#dfn-literal) matches a datatype if the [literal](https://www.w3.org/TR/shacl/#dfn-literal)'s datatype has the same [IRI](https://www.w3.org/TR/shacl/#dfn-iri) and, for the datatypes supported by SPARQL 1.1, is not an [ill-typed](https://www.w3.org/TR/rdf11-concepts#section-Graph-Literal) literal.

so some XSD datatypes are special.

TQ SHACL (java) generates a violation for `"test"^^xsd:date` (xsd:date isn't a SPARQL 1.1 datatype).
TQ SHACL gives a violation on "test"^^xsd:int.

**Reply by hmottestad** [reply]
Mischievous of you Andy to edit the spec while no one was watching. I could have sworn it didn't say anything about ill-typed literals when I was reading it. I guess that does make this a bug.

**Reply by afs** [reply]
:smile: Clearly, you are watching!

I didn't write any of the spec.

Test [core/property/datatype-ill-formed](https://github.com/w3c/data-shapes/blob/gh-pages/data-shapes-test-suite/tests/core/property/datatype-ill-formed.ttl) in the suite. It might have been added after the others - it has a different naming style.

Personally, i think it would better to also say something like "the processor should report any ill-defined literals it can identify".
